source/git/gitsource.go

```diff
@@ -232,6 +238,70 @@ func (gs *gitSourceHandler) getAuthToken(ctx context.Context, g session.Group) e
 	})
 }
 
+func (gs *gitSourceHandler) mountSSHAuthSock(ctx context.Context, g session.Group) (string, func() error, error) {
+	sshID := "default"
+	var caller session.Caller
+	err := gs.sm.Any(ctx, g, func(ctx context.Context, _ string, c session.Caller) error {
+		if err := sshforward.CheckSSHID(ctx, c, sshID); err != nil {
+			if st, ok := status.FromError(err); ok && st.Code() == codes.Unimplemented {
+				return errors.Errorf("no SSH key %q forwarded from the client", sshID)
+			}
+
+			return err
+		}
+		caller = c
+		return nil
+	})
+	if err != nil {
+		return "", nil, err
+	}
+
+	usr, err := user.Current()
+	if err != nil {
+		return "", nil, err
+	}
+
+	// best effor, default to root
+	uid, _ := strconv.Atoi(usr.Uid)
+	gid, _ := strconv.Atoi(usr.Uid)
+
+	sock, cleanup, err := sshforward.MountSSHSocket(ctx, caller, sshforward.SocketOpt{
+		ID:   sshID,
+		UID:  uid,
+		GID:  gid,
+		Mode: 0700,
+	})
+	if err != nil {
+		return "", nil, err
+	}
+
+	return sock, cleanup, nil
+}
+
+func (gs *gitSourceHandler) mountKnownHosts(ctx context.Context) (string, func() error, error) {
+	if gs.src.KnownSSHHosts == "" {
+		return "", nil, errors.Errorf("no configured known hosts forwarded from the client")
```

Comment:
Should figure out a way that the default behavior continues to work. Possible options are to disable hosts check if not set, perform a key scan from the client side, hardcode some builtin keys for known servers(github, bitbucket) on client side.

Reply:
`gs.mountKnownHosts()` is currently only being called if `gs.src.KnownSSHHosts != ""`, so this error won't be triggered. Which will then cause the `~/.ssh/known_hosts` file to be used if it exists.

Reply:
We shouldn't depend on any files from the host config in any case(and eventually move away from calling git binary in host context at all). If it is empty we should disable hosts check. And might be better if default from the client is not empty (eg. https://github.com/szuecs/go-ssh-keyscan/blob/master/go-ssh-keyscan.go showed up on a quick google search).

Reply:
Isn't that what it currently depends on? I didn't want to break the existing functionality. If you would rather I disable ssh host checking via `-o "StrictHostKeyChecking no"` when no known hosts are given we can do that, but it seems like an unsafe default option.

Upon a quick test of the linked go-ssh-keyscan, that doesn't seem any different than the regular ssh-keyscan, which is akin to just auto-accepting any new keys (or just disabling StrictHostKeyChecking).

I think auto-defaulting to adding in some known keys for github or bitbucket is a safer option.

Reply:
I think it is fine to break that as it was never supposed to work like this. If you want you can leave an environment variable to override it.

> Upon a quick test of the linked go-ssh-keyscan, that doesn't seem any different than the regular ssh-keyscan, which is akin to just auto-accepting any new keys (or just disabling StrictHostKeyChecking).

Yes, only thing it does is verify the key from two separate locations. I guess another thing that could be done is check if the key has not changed if a repo is reused in subsequent builds but that definitely doesn't need to be in this PR. I'm fine with disabling by default, except known hosts.

Reply:
That sounds fine, I added in that logic into a new function called `getGitSSHCommand()`.

Reply:
> hardcode some builtin keys for known servers(github, bitbucket) on client side.

Reply:
> hardcode some builtin keys for known servers(github, bitbucket) on client side.

Do you still want to do this. Can be a follow-up PR